lib/puppet/util/windows/adsi.rb

```diff
@@ -227,12 +236,62 @@ def password_is?(password)
       self.class.logon(name, password)
     end
 
-    def add_flag(flag_name, value)
-      flag = native_user.Get(flag_name) rescue 0
+    ADSI_FLAGS = {
+        script: 0x0001,
+        accountdisable: 0x0002,
+        homedir_required: 0x0008,
+        lockout: 0x0010,
+        passwd_notreqd: 0x0020,
+        passwd_cant_change: 0x0040,
+        encrypted_text_pwd_allowed: 0x0080,
+        temp_duplicate_account: 0x0100,
+        normal_account: 0x0200,
+        interdomain_trust_account: 0x0800,
+        workstation_trust_account: 0x1000,
+        server_trust_account: 0x2000,
+        dont_expire_password: 0x10000,
+        mns_logon_account: 0x20000,
+        smartcard_required: 0x40000,
+        trusted_for_delegation: 0x80000,
+        not_delegated: 0x100000,
+        use_des_key_only: 0x200000,
+        dont_req_preauth: 0x400000,
+        password_expired: 0x800000,
+        trusted_to_auth_for_delegation: 0x1000000,
+        partial_secrets_account: 0x04000000,
+    }
+
+    def flags
```

Comment:
I also needed to add some code for the ADSI flags while doing PUP-5216 -- here: https://github.com/ekinanp/puppet/commit/PUP-5216#diff-6ff32d97083beefa7a033714c7b30516R309 which is a bit different than how you have it set-up. We should probably agree on a consistent format for this code to avoid possible merge conflicts.

My thoughts are:
* I think the flag names should conform to what's in the Microsoft docs -- here: https://docs.microsoft.com/en-us/windows/desktop/api/iads/ne-iads-ads_user_flag. So in this case, it would be good to use my variable names. That's also how I've seen e.g. FFI constants defined in the Windows codebase.
* I like using a hash to specify all of the flags. In fact, I revised mine to use a hash after I saw yours earlier :).
* I think it might be better to use `set` instead of `add`, `unset` instead of `remove`, and something like `flag_set?` instead of `has_flag?` just because I think for bitwise flags, `set`/`unset` are the common language instead of `add`/`remove`. Don't have a strong opinion on this one though.
* My reason for using `set_userflags` instead of `set_flags` is b/c `userflags` is the ADS property so I was thinking since ADSI::User is meant to connect to the IADsUser, it would be good to keep those consistent. But I am fine with `flags` too.
* I have my validation code in the `set/unset` instead of `userflags_set?` because:
** `userflags_set?` will return `false` for an unrecognized flag, which is still correct b/c the latter isn't set.
** `set/unset`  is where we can manipulate `userflags` so it seems like that's where the validation matters.
Also stole the validation idea from you. 

For context, the reason I have `*flags` in `set/unset` instead of `flag` is b/c I need to set multiple userflags for PUP-5216 and specifying multiple flags avoids multiple commits. I think it would be good to keep that interface.

Reply:
What do you think?